cpp/src/parquet/page_index.cc

```diff
@@ -184,8 +185,241 @@ class OffsetIndexImpl : public OffsetIndex {
+  PageIndexReaderImpl(::arrow::io::RandomAccessFile* input,
+                      std::shared_ptr<FileMetaData> file_metadata,
+                      const ReaderProperties& properties,
+                      std::shared_ptr<InternalFileDecryptor> file_decryptor)
+      : input_(input),
+        file_metadata_(std::move(file_metadata)),
+        properties_(properties),
+        file_decryptor_(std::move(file_decryptor)) {}
+
+  ::arrow::Result<std::shared_ptr<RowGroupPageIndexReader>> RowGroup(int i) override {
+    if (i < 0 || i >= file_metadata_->num_row_groups()) {
+      return ::arrow::Status::IndexError("Invalid row group ordinal {}", i);
+    }
+    return std::make_shared<RowGroupPageIndexReaderImpl>(
+        input_, file_metadata_->RowGroup(i), properties_, i, file_decryptor_);
+  }
+
+  void WillNeed(const std::vector<int32_t>& row_group_indices, bool need_column_index,
+                bool need_offset_index) override {
+    std::vector<::arrow::io::ReadRange> read_ranges;
+    for (int32_t row_group_ordinal : row_group_indices) {
+      int64_t column_index_offset;
+      int64_t column_index_size;
+      int64_t offset_index_offset;
+      int64_t offset_index_size;
+      bool has_column_index;
+      bool has_offset_index;
+      PageIndexReader::DeterminePageIndexRangesInRowGroup(
+          *file_metadata_->RowGroup(row_group_ordinal), &column_index_offset,
+          &column_index_size, &offset_index_offset, &offset_index_size, &has_column_index,
+          &has_offset_index);
+      if (need_column_index && has_column_index) {
+        read_ranges.emplace_back(
+            ::arrow::io::ReadRange{column_index_offset, column_index_size});
+      }
+      if (need_offset_index && has_offset_index) {
+        read_ranges.emplace_back(
+            ::arrow::io::ReadRange{offset_index_offset, offset_index_size});
+      }
+    }
+    PARQUET_IGNORE_NOT_OK(input_->WillNeed(read_ranges));
+  }
+
+  void WillNotNeed(const std::vector<int32_t>& row_group_indices) override {
+    // No-op for now.
+  }
+
+ private:
+  /// The input stream that can perform random read.
+  ::arrow::io::RandomAccessFile* input_;
+
+  /// The file metadata to get row group metadata.
+  std::shared_ptr<FileMetaData> file_metadata_;
+
+  /// Reader properties used to deserialize thrift object.
+  const ReaderProperties& properties_;
+
+  /// File-level decrypter.
+  std::shared_ptr<InternalFileDecryptor> file_decryptor_;
+};
+
 }  // namespace
 
+void PageIndexReader::DeterminePageIndexRangesInRowGroup(
+    const RowGroupMetaData& row_group_metadata, int64_t* column_index_start,
+    int64_t* column_index_size, int64_t* offset_index_start, int64_t* offset_index_size,
+    bool* has_column_index, bool* has_offset_index) {
+  int64_t ci_start = std::numeric_limits<int64_t>::max();
+  int64_t oi_start = std::numeric_limits<int64_t>::max();
+  int64_t ci_end = -1;
+  int64_t oi_end = -1;
+  for (int i = 0; i < row_group_metadata.num_columns(); ++i) {
```

Comment:
should there be a version of this method that also takes in the columns selected?

Reply:
That is feasible. However, it would complicate the implementation if user has provided selected columns and then requests to read page index of a non-selected column.

Reply:
it feels like this optimization could be worth it, especially in the tail.  I've observed parquet files with O(1000) of columns, it isn't clear to me if page indexes are generally enabled for all columns or not though.  I think if we provided this method, then we simply document behavior as throwing an exception if the request wasn't requested, but it seems fine todo as a follow-up if we want it.

Reply:
I agree that if a parquet file contains many columns, it would be painful to coalesce them into a single I/O chunk. In our use case to optimize reading file tail, we simply cache the coalesced chunk of all columns for repeated read because it is difficult to predict which columns will be accessed next. I can add a function to optionally specify requested columns and by default all columns are requested.

Reply:
@pitrou Do you have the chance to take a look into this? I'd like to hear your advice on the proposed `WillNeed` and `WillNotNeed` interfaces we have discussed here. Especially the granularity of row group or column.

Reply:
I only took a brief look but I think the current granularity of row group + column index is sufficient. That was the scenario we benchmarked (selecting columns out of a file) and we didn't need anything finer-grained. (IIRC, we ended up with only a single row group per file, too.)